core/dbt/clients/agate_helper.py

```diff
@@ -84,21 +93,36 @@ def table_from_data(data, column_names: Iterable[str]) -> agate.Table:
         table = agate.Table.from_object(data, column_types=DEFAULT_TYPE_TESTER)
         return table.select(column_names)
 
-
 def table_from_data_flat(data, column_names: Iterable[str]) -> agate.Table:
-    "Convert list of dictionaries into an Agate table"
+    """
+    Convert a list of dictionaries into an Agate table. This method does not
+    coerce string values into more specific types (eg. '005' will not be
+    coerced to '5'). Additionally, this method does not coerce values to
+    None (eg. '' or 'null' will retain their string literal representations).
+    """
 
     rows = []
+    text_only_columns = set()
     for _row in data:
         row = []
-        for value in list(_row.values()):
+        for col_name in column_names:
+            value = _row[col_name]
             if isinstance(value, (dict, list, tuple)):
-                row.append(json.dumps(value, cls=dbt.utils.JSONEncoder))
-            else:
-                row.append(value)
+                # Represent container types as json strings
+                value = json.dumps(value, cls=dbt.utils.JSONEncoder)
+                text_only_columns.add(col_name)
```

Comment:
I was originally just going to peek at the first row in the result set, but it's possible that the value in a string column returned by the database could be `NULL`. This implementation looks at all of the values in the result set, which we're actually already doing anyway to json-encode dict/list/tuple types.